### ISSUE TITLE
Error HRESULT E_FAIL with iTunes

### ISSUE THREAD

**Issue by areeves89** [dlrudie/Snip#202]
I updated to the latest version of iTunes for Windows 10 (version 12.7.0.166), but since doing so it doesn't seem to want to play nicely with Snip (version 6.9.1). Whenever I open Snip, it opens iTunes as well, but immediately after iTunes opens I get the following error:

_Please exit iTunes and reselect iTunes from within Snip.

Error HRESULT E_FAIL has been returned from a call to a COM component._ 

iTunes seems to function fine, but Snip is unable to pick up song info from iTunes. I've tried closing and selecting Snip->iTunes to re-launch iTunes as instructed in the error message, but the same error comes up each time. I've experimented with setting Snip to another service (Spotify, VLC, etc.), opening iTunes separately, and then setting it back to iTunes, and I've also uninstalled and re-installed, but despite my best efforts the error message comes up each time. Has anyone else seen this kind of thing before?

**Comment by cheeseluigi** [comment]
Yep I'm also encountering this issue, it appears the new iTunes update broke snip's compatibility. If anyone has a dl for the previous version of iTunes that'd be appreciated until this is fixed.

**Comment by Kurtessis** [comment]
https://support.apple.com/downloads/itunes

**Comment by dlrudie** [comment]
I’ve been looking into it. Have not found a solution yet. Apple broke compatibility with their own library.

**Comment by dlrudie** [comment]
So it's not the COM library that's broken. Internal events within the COM library have changed. There's zero documentation on this so I'm stuck playing around with it until I find something that works.

**Comment by dlrudie** [comment]
I've been trying to figure out a solution but am still stuck.

If I don't create any events I can link against the COM library and the rest of the stuff works. But when trying to create an event it has the E_FAIL error.

this.iTunesApplication.OnPlayerPlayEvent += new _IiTunesEvents_OnPlayerPlayEventEventHandler(this.App_OnPlayerPlayEvent);

The problem seems to be that there is no longer these event handlers. It will still build just fine but when trying to use them it fails.

I've tried digging around other things within the library but nothing comes up. They either changed them or removed them internally.

**Comment by VickGE** [comment]
Hey, Rainmeter did fix their iTunes problem, it's surely a different, but you could contact their devs to see how they did it.

**Comment by 1215drew** [comment]
It actually looks like it could be very related. Here's the pull request from the rainmeter project: https://github.com/rainmeter/rainmeter/pull/125

**Comment by CyberStar1** [comment]
Any way to fix this yet?

**Comment by markspolakovs** [comment]
@CyberStar1

> There's zero documentation on this so I'm stuck playing around with it until I find something that works.

> They either changed them or removed them internally.

**TL;DR**: no

**Comment by VickGE** [comment]
@markspolakovs Other programs like Rainmeter found a solution to this problem, so it is possible to fix now.

**Comment by markspolakovs** [comment]
I've created a hotfix for the issues. Until #238 is merged, you can download my hotfixed version at https://github.com/markspolakovs/Snip/releases/tag/v6.9.1-hotfix.1.

Please note that this is probably broken and unsupported - do not report any issues here!

**Comment by kmaehashi** [comment]
Hi, I'm a developer of @LyricsMaster and having the same issue.
This is just a FYI for developers, not a solution nor a workaround: it seems that iTunes is not dropping COM event support, but rather it was caused by a configuration error of iTunes.
iTunes v12.7 specifies `PSOAInterface` as a proxy/stub while it was `PSDispatch` in v12.6.
I confirmed that by changing the registry value `HKEY_CLASSES_ROOT\Interface\{5846EB78-317E-4B6F-B0C3-11EE8C8FEEF2}\ProxyStubClsid32` from `{00020424-0000-0000-C000-000000000046}` (CLSID of `PSOAInterface`) to `{00020420-0000-0000-C000-000000000046}` (`PSDispatch`), events are dispatched as it was in v12.6.

**Comment by dlrudie** [comment]
Thanks for the insight! I wonder if there's a way to override the CLSID in code. May be worth reaching out to Apple somehow to see what they can do.

**Comment by kmaehashi** [comment]
I've already reported this to Apple on Dec 9th via https://bugreport.apple.com, but got response so far...

**Comment by dlrudie** [comment]
See #223 for hotfix for now.

**Comment by dlrudie** [comment]
I just updated to iTunes 12.7.4 and they appear to have switched back to PSDispatch. Snip is working perfectly fine again with iTunes.
